src/daemon/daemon.cpp

```diff
@@ -2217,40 +2217,32 @@ void mp::Daemon::create_vm(const CreateRequest* request, grpc::ServerWriter<Crea
             reply.set_create_message("Configuring " + name);
             server->Write(reply);
 
-            // To generate the network interfaces, we need to check first for repetition the MAC addresses
-            // which were specified by the user, and add them to the set of new MAC addresses. If it happens
-            // that we have repeated addresses, fail. Only after doing this we will be able to generate the
-            // rest of the addresses.
-            for (const auto& iface : checked_args.extra_interfaces)
-                if (!iface.mac_address.empty() &&
-                    (allocated_mac_addrs.find(iface.mac_address) != allocated_mac_addrs.end() ||
-                     !added_mac_addresses.insert(iface.mac_address).second))
+            // Generate mac if empty, check for repetition otherwise; let the backend re-interpret the id
+            for (auto& iface : checked_args.extra_interfaces)
+            {
+                iface.id = config->factory->interface_id(iface.id);
+
+                if (iface.mac_address.empty())
+                    iface.mac_address = generate_unused_mac_address(added_mac_addresses);
+                else if (allocated_mac_addrs.find(iface.mac_address) != allocated_mac_addrs.end() ||
+                         !added_mac_addresses.insert(iface.mac_address).second)
                     throw std::runtime_error(fmt::format("Repeated MAC address {}", iface.mac_address));
+            }
 
             // Generate a default network interface.
             mp::NetworkInterface default_interface{"default", generate_unused_mac_address(added_mac_addresses), true};
 
-            // Generate MAC addresses for the interfaces on which it was not specified and put the modified
-            // interfaces in a new vector.
-            std::vector<mp::NetworkInterface> extra_interfaces;
-            for (const auto& iface : checked_args.extra_interfaces)
-            {
-                extra_interfaces.push_back(mp::NetworkInterface{
-                    config->factory->interface_id(iface.id),
-                    iface.mac_address.empty() ? generate_unused_mac_address(added_mac_addresses) : iface.mac_address,
-                    iface.auto_mode});
-            }
-
```

Comment:
There is a reason to check first the specified MAC address and generating addresses afterwards: all the addresses specified by the user must be added to `added_mac_addresses` first, so they are taken into account when generating new addresses. I'd like to keep that behavior.

OTOH, the non-const version is better, we can avoid copying the vector.

Reply:
Right, good point! I missed that. I would still prefer to condense the code somehow... Let me see if I can come up with something.

Reply:
@luis4a0 I have just pushed a fix, WDYT?

Indeed, I don't see a way to escape the two loops. I thought to [partition](https://en.cppreference.com/w/cpp/algorithm/partition), but I didn't want to change the order that the user requested (and of course that internally would also loop).

Reply:
Yes, partitioning is a possibility but it's better to keep the users order. The code you pushed works and it's clear enough, let's leave it like that.